pkg/cosign/verify_blob.go

```diff
@@ -0,0 +1,223 @@
+	"github.com/sigstore/rekor/pkg/types"
+	hashedrekord "github.com/sigstore/rekor/pkg/types/hashedrekord/v0.0.1"
+	rekord "github.com/sigstore/rekor/pkg/types/rekord/v0.0.1"
+	"github.com/sigstore/sigstore/pkg/cryptoutils"
+	"github.com/sigstore/sigstore/pkg/signature"
+	"github.com/sigstore/sigstore/pkg/signature/dsse"
+)
+
+type blobSignature struct {
+	payload []byte
+	b64sig  string
+	bundle  *bundle.RekorBundle
+	v1.Layer
+}
+
+func newBundleSignature(blobBytes []byte, b64sig string, bundle *bundle.RekorBundle) (*blobSignature, error) {
+	if blobBytes == nil {
+		return nil, errors.New("blobBytes must be non nil")
+	}
+	if b64sig == "" {
+		return nil, errors.New("b64sig must be non empty string")
+	}
+	return &blobSignature{
+		payload: blobBytes,
+		b64sig:  b64sig,
+		bundle:  bundle,
+	}, nil
+}
+
+func (s *blobSignature) Annotations() (map[string]string, error) {
+	return nil, nil
+}
+
+func (s *blobSignature) Payload() ([]byte, error) {
+	return s.payload, nil
+}
+
+func (s *blobSignature) Base64Signature() (string, error) {
+	return s.b64sig, nil
+}
+
+func (s *blobSignature) Cert() (*x509.Certificate, error) {
+	return nil, errors.New("no cert in blobSignature")
+}
+
+func (s *blobSignature) Chain() ([]*x509.Certificate, error) {
+	return nil, errors.New("no cert chain in blobSignature")
+}
+
+func (s *blobSignature) Bundle() (*bundle.RekorBundle, error) {
+	return s.bundle, nil
+}
+
+// VerifyBlobSignature verifies a signature
+func VerifyBlobSignature(ctx context.Context, blobBytes []byte, b64sig string, bundle *bundle.RekorBundle, co *CheckOpts) (checkedSignatures []oci.Signature, bundleVerified bool, err error) {
+	sig, err := newBundleSignature(blobBytes, b64sig, bundle)
+	if err != nil {
+		return checkedSignatures, bundleVerified, errors.Wrap(err, "failed to create bundle signature")
+	}
+
+	verifiers := []signature.Verifier{}
+	if co.SigVerifier != nil {
+		verifiers = append(verifiers, co.SigVerifier)
+	} else {
+		uuids, err := FindTLogEntriesByPayload(ctx, co.RekorClient, blobBytes)
+		if err != nil {
+			return checkedSignatures, bundleVerified, errors.Wrap(err, "failed to get tlog entries by payload")
+		}
+		if len(uuids) == 0 {
+			return checkedSignatures, bundleVerified, errors.New("could not find a tlog entry for provided blob")
+		}
+		for _, u := range uuids {
+			tlogEntry, err := GetTlogEntry(ctx, co.RekorClient, u)
+			if err != nil {
+				continue
+			}
+			certs, err := extractCerts(tlogEntry)
+			if err != nil {
+				continue
+			}
+			cert := certs[0]
+			verifier, err := ValidateAndUnpackCert(cert, co)
+			if err != nil {
+				continue
+			}
+			// Use the DSSE verifier if the payload is a DSSE with the In-Toto format.
+			if isIntotoDSSE(blobBytes) {
+				verifier = dsse.WrapVerifier(verifier)
+			}
+			verifiers = append(verifiers, verifier)
+		}
+	}
+
+	var validSigExists bool
+	for _, verifier := range verifiers {
+		if err := verifyOCISignature(ctx, verifier, sig); err != nil {
```

Comment:
What happens to intoto attestations?

Reply:
Thank you for your reviewing! The intoto DSSE verifier will only be available in keyless verification mode with my current codes, and this is simply a mistake.
I will fix this.